### DISCUSSION TITLE
Input widget best practices

### DISCUSSION THREAD

**Discussion by akhudek** [fltk-rs/fltk-rs#1376] [Q&A]
I've noticed that the standard callback on the Input widget doesn't trigger except when you focus another input. While I could instead use the change callback to trigger updates to internal data models, it would be nicer if the Input widget behaved more like other toolkits. 

E.g. if I click on some other element it would lose focus and trigger the standard callback. Or perhaps hitting enter or escape would also do the same. I have an idea how to implement these things, but wondered if someone had already tackled this problem.

**Comment by MoAlyousef** [comment]
You can use input.set_trigger(CallbackTrigger::EnterKeyAlways) to get the callback when the user presses enter. 
I’m not sure whether losing focus should trigger the callback would be favorable, but you can bitor Release to EnterKeyAlways to get that behavior. 

I’m open to other views as well.

**Reply by akhudek** [reply]
Thanks, I'm using set_trigger currently, though the issue is more that the input widget doesn't release/unfocus when you click away. It will if you click into another text input, but if you click a button, dropdown, or just click on a frame in the app the cursor remains in the input and it still has focus. 

The usual behaviour in apps in general is to defocus the input if the user interacts with another widget/element in any way. Ideally I'd like fltk to behave this way too and thought there might be some easy way to achieve it aside from creating a custom input widget that replaces handle or similar. 

You can see this behaviour in the temp_converter.rs example. Clicking either box defocuses the other, but clicking on a blank part of the app doesn't defocus.

**Reply by MoAlyousef** [reply]
Ok I get what you mean now. You might be able to ask the fltk devs in the [fltk mailing list](https://groups.google.com/g/fltkgeneral) if they would consider such a change in behavior.
